Report List Api

MinieMoe opened the pull request:
Implement API call for frontend to make LIST request and fetch all reports from the backend.
![image](https://github.com/CodeDayLabs311/cl311/assets/93412982/7c8f44f0-a513-4cfb-af93-30fc12ac7fab)

**Testing:**
Since I don't have the create report page to create item to put testing data in, I made a function mocking DynamoDB sending all reports to the backend. Next step is to get it merged with Sophie and Zhanping's works so it connects to DynamoDB.

MinieMoe committed:
Report list page

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Add more pages that I forgot to add

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
Change name and add accessiblity for screenreader

MinieMoe committed:
adding new files to build api call

MinieMoe committed:
adding new files to build api call

MinieMoe committed:
adding new files to build api call

MinieMoe committed:
adding new files to build api call

MinieMoe committed:
Adding files for API calls to backend for fetching all reports

MinieMoe committed:
Adding files for API calls to backend for fetching all reports

MinieMoe committed:
Adding files for API calls to backend for fetching all reports

MinieMoe committed:
Adding files for API calls to backend for fetching all reports

MinieMoe committed:
Add listReport API

MinieMoe committed:
Add listReport API

MinieMoe committed:
Add listReport API

MinieMoe committed:
Add hook to fetch all reports

MinieMoe committed:
Add hook to fetch all reports

MinieMoe committed:
Add hook to fetch all reports

MinieMoe committed:
Integrate API call to report list page

MinieMoe committed:
Integrate API call to report list page

MinieMoe committed:
Integrate API call to report list page

MinieMoe committed:
Fixing reportListing response

MinieMoe committed:
Fixing reportListing response

MinieMoe committed:
Fixing reportListing response

MinieMoe committed:
Mock listing API with fake data. It worksgit status

MinieMoe committed:
Mock listing API with fake data. It worksgit status

MinieMoe committed:
Mock listing API with fake data. It worksgit status

MinieMoe committed:
Integrate mock data into table

MinieMoe committed:
Integrate mock data into table

MinieMoe committed:
Integrate mock data into table

MinieMoe commented on app/src/pages/api/report/list/index.ts:
```diff
@@ -0,0 +1,94 @@
+/** NOTE: function to mock DB lisitng because I don't have access to create report page or permission to create new item in AWS
+    TODO: will need to switch to real DB function eventually
+*/
+const mockListReportsFromDb = (): { reports: IReport[]; paginationToken: string | undefined } => {
+    return {
+        reports: [
+            {
+                reportId: 'report1',
+                name: 'John Doe',
+                emailAddress: 'john.doe@example.com',
+                phoneNumber: '123-456-7890',
+                reportCategory: ['Illegal Dumping'],
+                address: '123 Main St',
```
@andreybutenko I made this mock function. Is it a good idea to put it here or should I put it somewhere else like a test or util folder? What do you think?

andreybutenko replied:
Good question! I think if we were using this data in multiple places or for unit tests, it should be in a test constants folder! But since this will be replaced shortly anyway, it is fine here until it is replaced with real db interaction :)

MinieMoe committed:
Adding button to link to edit and view page

MinieMoe committed:
Adding button to link to edit and view page

MinieMoe committed:
Adding button to link to edit and view page

MinieMoe committed:
Merge branch 'min-reportListPage' of https://github.com/CodeDayLabs311/cl311 into min-reportListPage

MinieMoe committed:
Update report list UI

MinieMoe committed:
Update report list UI

MinieMoe committed:
Add more report statuses

MinieMoe committed:
Add more report statuses

MinieMoe committed:
Form validation (#4)

Form validation should work now for both when you create a new message
and edit a pre-existing message

---------

Co-authored-by: Min <nguye384@wwu.edu>

andreybutenko commented on app/src/components/report/Table.tsx:
```diff
@@ -0,0 +1,164 @@
+            headerName: 'Issues',
+            //TODO: change this to something else, because this is an array of issues
+            type: 'singleSelect',
+            valueOptions: ['Illegal dump', 'Clogged drains', 'Other'],
```
Can these be moved to an enum so it's easy to keep values in sync between the backend and the frontend?

andreybutenko commented on app/src/components/report/Table.tsx:
```diff
@@ -0,0 +1,164 @@
+        {
+            field: 'reportCategory',
+            headerName: 'Issues',
+            //TODO: change this to something else, because this is an array of issues
```
Would it make sense for each report to have a single category?

MinieMoe replied:
Sophie said that a report can have multiple issues categories and defined the 'Issues' column as a list of string in the DB. I'm not sure about this. What do you think?

andreybutenko commented on app/src/components/report/Table.tsx:
```diff
@@ -0,0 +1,164 @@
+            headerName: 'Status',
+            //this type is for filtering and editing
+            type: 'singleSelect',
+            valueOptions: ['Reported', 'Pending', 'Dispatched', 'Resolved'],
```
Should these move to an enum?

andreybutenko commented on app/src/components/report/Table.tsx:
```diff
@@ -0,0 +1,164 @@
+                        <div>
+                            {/* First row item */}
+                            {cellValues.value.map((issue: any) => (
+                                <p style={{margin:0}}>{issue}</p>
```
Nit: should be styled as `{{ margin: 0 }}`

(install prettier in your IDE for auto formatting)

andreybutenko commented on app/src/hooks/report/useReports.ts:
```diff
@@ -0,0 +1,46 @@
+export const useReports = (): UseReportsResult => {
+    const reportClient = useReportClient();
+
+    const [reports, setReports] = useState<IReport[] | undefined>([]);
```
If `undefined` indicates that it is loading, should the initial value be `undefined` instead of `[]`?

andreybutenko commented on app/src/models/clients/IReportClient.ts:
```diff
@@ -0,0 +1,15 @@
+import { IReport } from '../data';
+
+/* Interface for all API calls of ReportClient */
+export interface IReportClient {
```
Nice setup!

andreybutenko commented on app/src/models/data/IReport.ts:
```diff
@@ -2,22 +2,16 @@
 export interface IReport {
```
Is there a reason to flatten `contactInformation`, `issueLocation`, etc? I don't have a strong preference either way, but I think it makes sense for these to be grouped :)

andreybutenko commented on app/src/utils/clients/ReportApiClient.ts:
```diff
@@ -0,0 +1,20 @@
+const LIST_REPORTS_ENDPOINT = '/api/report/list'
+
+/** Implementation of IReportClient interface */
+export class ReportApiClient implements IReportClient {
```
Nice setup here too!

MinieMoe committed:
Change the initial state of the reports

MinieMoe committed:
Change the initial state of the reports

MinieMoe committed:
Change the initial state of the reports

MinieMoe committed:
Address changes from PR feedbacks

MinieMoe committed:
Address changes from PR feedbacks

SophieNguyen113 committed:
DB Connection and Full Stack Create a Report (#10)

**1. DB Connection between backend and DynamoDB:** Created
ReportDbClient.ts to make these functions: Create report, Get report,
List reports, List reports by category (Min said she didn't need this
function since MUI has a pretty cool filter already, I still decided to
keep this though), Put report, and Unmarshal reports to/from DynamoDB.

**2. Full Stack (Frontend + Backend) Create a Report (Zhanping is still
working on the UI):** Implemented createReport and getReport functions
in IReportClient.ts, the interface for report clients. In
ReportApiClient.ts, built create and get endpoints to interact with
report API, etc.

**3. Demo:**

![Create a
Report](https://github.com/CodeDayLabs311/cl311/assets/123401167/2661b138-a05c-4418-9b99-619d8fada5ba)

In this demo, I showed how to create a report. 
I could fill in all the fields. 
In the 'Report Category', I could check/uncheck multiple categories
(type="checkbox").
The 'Status Of Report' was initially set up as 'Submitted', I could only
choose 1 option (type="radio"), I could choose another option (In
Progress, Completed, On Hold, Rejected) if I would like to.
The 'Date Time Of Submission' was read-only, I couldn't change this
field.
When I clicked on the "Create" button, I could store all the data in
DynamoDB. _**The proof was that you could see the ID that I highlighted
on the GIF demo.**_

_**Explain the 'Page Not Found':**_ Since this PR is just about
'Creating a Report', when you click on the "Create" button, it's
supposed to show up the 'Viewing Page', which is the next PR,
Viewing-Page-Full-Stack.

Thanks :>

MinieMoe committed:
Add new Query permission to DB (#16)

`actions: ['dynamodb:Scan', 'dynamodb:GetItem', 'dynamodb:PutItem',
'dynamodb:Query'],`

I'm adding new permission to DynamoDB so we can do query stuff with it

MinieMoe committed:
apply changes from reportListPage

MinieMoe committed:
apply changes from reportListPage

MinieMoe committed:
Fixing diverge in local and remote branch

MinieMoe committed:
Rebased from main and fixed merge

MinieMoe committed:
Merge branch 'min-reportListAPI' of https://github.com/CodeDayLabs311/cl311 into min-reportListAPI

MinieMoe committed:
Fixed merge

MinieMoe committed:
Fixing incompatible type

MinieMoe committed:
Finishing up rebasing

MinieMoe committed:
Remove reportId from printed

MinieMoe committed:
Fix enum
